modules/library/data/src/main/java/org/geotools/data/memory/MemoryFeatureReader.java

```diff
@@ -30,20 +34,27 @@
 import org.opengis.feature.simple.SimpleFeatureType;
 
 public class MemoryFeatureReader implements FeatureReader<SimpleFeatureType, SimpleFeature>{
-    ContentState state;
+
     SimpleFeatureType featureType;
     Iterator<SimpleFeature> iterator;
 
     public MemoryFeatureReader(ContentState state, Query query) throws IOException {
-        this.state = state;
         featureType = state.getFeatureType();
-        String typeName = getFeatureType().getTypeName();
-        MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
-        iterator = store.features(typeName).values().iterator();
+
+        final MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
+        final Map<String, SimpleFeature> features = store.features(featureType.getTypeName());
```

Comment:
This is driving me bonkers, the point is to have the Map<String,SimpleFeature> on each Entry as the canonical truth. And a diff map on each entry that could be used by readers/writers.

I looks like this transition to contentdatastore was done with a minimal amount of code change, introducing all kinds of opportunity for error.